plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsStepContextTest.java

```diff
@@ -181,4 +182,34 @@ static ClosureCounter get() {
             return ExtensionList.lookupSingleton(ClosureCounter.class);
         }
     }
+
+    @Test public void refersToCycle() throws Exception {
+        logger.record(CpsStepContext.class, Level.ALL);
+        var p = r.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("refersToCycle()", true));
+        r.assertLogNotContains("StackOverflowError", r.buildAndAssertStatus(Result.UNSTABLE, p));
```

Comment:
Interestingly, without the fix, the build is still `UNSTABLE` rather than `FAILURE`, apparently because the `StackOverflowError` is thrown out of the step body and thus becomes the build-terminating error yet the status has already been set by the `FlowInterruptedException`. Still, the stack trace is printed.

If you

```diff
-ctx.onFailure(t1);
+throw t1;
```

then you still get the stack overflow, but it is coming from the pool thread used by `SynchronousNonBlockingStepExecution` so it just gets basically ignored.